docs/proposals/coap-core.md

```diff
@@ -0,0 +1,74 @@
+    ```
+    </sensors/temp>;rt="oic.r.temperature";if="sensor",
+    </sensors/light>;rt="oic.r.light.brightness";if="sensor"
+    ```
+
+    The example is stating that the device has 2 REST resources, `/sensors/temp` and `/sensors/light` which are of type `oic.r.temperature` and `oic.r.light.brightness` respectively. `rt` values are defined in [IANA](https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#rt-link-target-att-value) to have same standardization, although vendor values can probably be used. Then `if` means that the interface description of the resource is of type `sensor`. Currently only resources with interface `sensor` are supported by the agent.
+3. For each device, the agent returns a `DiscoveryResult` which will have the following properties:
+    ```
+    COAP_IP:                 192.168.1.126
+    COAP_RESOURCE_TYPES:     oic.r.temperature,oic.r.light.brightness
+    oic.r.light.brightness:  /sensors/light
+    oic.r.temperature:       /sensors/temp
+    ```
+    `COAP_IP` and `COAP_RESOURCE_TYPES` are static and available for each Instance. `oic.r.light.brightness` and `oic.r.temperature` are dynamic, based on the discovered resources. By doing so, the cluster can look for a resource (e.g. temperature measurements) by searching for Instances which support the `oic.r.temperature` resource. 
+4. An Akri Broker is provisioned for each Istance. The Broker has the following environment variables based on the previous properties and can be reached via its associated service:
+
+  ```
+  COAP_IP=192.168.1.126
+  oic.r.temperature=/sensors/temp
+  oic.r.light.brightness=/sensors/light
+  COAP_RESOURCE_TYPES=oic.r.temperature,oic.r.light.brightness
+  ```
+
+  The Broker acts as a HTTP to CoAP Proxy. It translates RESTful HTTP requests into RESTful CoAP requests and viceversa for the response. "Cross-Protocol Proxying between CoAP and HTTP" is defined in section 10 of RFC 7252. Currently, the Broker forwards only GET requests.
+  
+  It's also in a good position to cache CoAP responses. "Unlike HTTP, the cacheability of CoAP responses does not depend on the request method, but it depends on the Response Code." Currently, the Broker caches 2.05 Content responses, which are returned if the device is not reachable during the connection.
+
+## Outstanding Questions
+
+- Is there a better way to store the discovered resources as Configuration in the cluster?
+
+I think the current implementation would need a controller to accept queries about available resources and return the broker which can communicate to the device. The device is listed as generic `akri.sh/coap-core-021dd7` resource on the node, which is pretty much useless. A better label would be `akri.sh/oic.r.temperature-021dd7`, that is the discovered resource. This would allow using the K8s controller for scheduling pods which need the resource.
```

Comment:
> are you thinking of adding support for filter in/out devices based on resource type, similar to Zeroconf

I can surely do it if there some use cases for it 🤔

About the rest of the point, I think that filtering to allow only a low fixed numbers of resource types could be limiting. If I were a platform like Azure, who wants to offer edge clusters with IoT devices as resources, I'd prefer supporting a large diverisity of resource types because that's the nature of IoT. 
I think filtering makes sense if the use case is a company is specialized in one kind of devices.

One of the ideas of my thesis, is building a platform which can offer local devices as services in edge computing. Just like today you can provision databases and other software as service in the cloud. But even in a small areas like a neighbourhood, you could have a large number of resource types. Also by filtering, that would mean that you would have two separate Configurations for temperature and brightness and thus two brokers. Having two Configurations doesn't look bad per se, whereas two brokers for each resource in each device feels more wrong. That number `brokers x types x devices` would be explosive I believe.

Reply:
I think the main question I have is whether each resource type should be returned as a separate `DiscoveryResult` [here](https://github.com/jiayihu/akri/blob/6aee1161d6eccdbc5a80051f38560f6b24692bbe/agent/src/protocols/coap_core/discovery_handler.rs#L49) and therefore gets a separate Instance (and could also get a separate broker). Currently, we have one broker deployment strategy of one broker per instance, but this could change in the future to where someone could request a core_coap temperature and light sensor per broker, but this does get weird, since that could result in getting a temperature sensor from one endpoint and a light sensor from another instead of the temperature and light sensors from the same endpoint. It would also be hard to manage device usage if multiple resources of the device are being exposed as instances.

I am not familiar with the use cases for coap_core, so I differ to you. The more we discuss it, I am falling in line with your thoughts of each service being a `DiscoveryResult` rather than breaking it down further by resource type. Thanks for being patient with me and helping me type through my thoughts.

As for filtering, I still think it could be useful. People might only want to discover services that expose a certain resource type, since their brokers may specifically be interested in that resource type. Say you have a broker that grabs the temperature reading from temperature sensors and serves it to some anomaly detection app, you'd want to ensure that the service/device it is grabbing the temperature from does in fact have a temperature resource. Then, in the Configuration, you could specify the filter and `discover` would simply only return devices that have that resource type.

Also, your thesis sounds cool. Do you intend to use existing tech, such as CoAP to expose the local devices as services? Do you intend to just use CoAP or discover the devices via multiple protocols?

Reply:
> someone could request a core_coap temperature and light sensor per broker, but this does get weird, since that could result in getting a temperature sensor from one endpoint and a light sensor from another instead of the temperature and light sensors from the same endpoint

Actually I don't think that becomes necessary weird. It would allow to virtualize the IoT resources. If you are asking for temperature and brightness, one could not care from which device it comes from. That would allow some virtualization benefits. For instance, we could use the brightness from another device if it fails in the current one. Since we're talking about edge, one may probably want to restrict the location of the resources though (e.g. "temperature and brightness in this room"). And if you restrict enough, the limit case would having the two resources from the same device. It's just an idea, maybe silly.

What I mean is that I think it depends on the use case. In other cases, I agree it does get weirds, e.g. if I need both camera recording and infrared recording from the same device.

The interesting point that has come to my mind while typing is that maybe we should discuss about what virtualization means for akri in the context of edge devices. Currently, akri allows to add device instances as resources and it's thus a virtualization of device instances. However, virtualizing resources instead of devices seems to be more natural to me if we think about the cloud.

For instance, let's be guided by how we expect users to use Akri instances. An application would never request a CoAP CoRE device, because that doesn't make sense. It would more likely happier to request a resource type, like temperature. How do we implement that with Akri? If we have an instance/broker per device, then we need to write a new controller which scans the list of device instances to find a result with the temperature sensor as metadata. This doesn't feel very natural to me. 

Is an instance per resource better? Maybe. Is also a broker per resource instance better? I don't think so because it would be an explosive overhead.

I think "the main question I have is whether each resource type should be returned as a separate DiscoveryResult" can be reduced further to what virtualization means for Akri and, even before that, what's the expected user experience for application developers. How will developers use CoRE resources in your opinion?

Do these questions make sense for the other protocols? Unfortunately I have no experience with the other protocols, although I saw OPC UA mentioned in some papers. My feeling is that maybe these questions haven't occurred before because many protocols assume what the underlying resource type is. E.g. an onvif instance is the same as a camera instance and thus requesting an onvif instance to Akri is fine. But if we think about it, the discovery protocol should be an implementation detail. If you need a camera recording, the cluster should be able to provision it independently from the protocol. If I have an OVIF camera, a udev camera and a CoAP CoRE camera, the three of them should be usable by the application. And this is also why my CoAP broker acts as a proxy converting between HTTP to CoAP and viceversa. Any application can request the CoAP resources without knowing about CoAP.

> Also, your thesis sounds cool. Do you intend to use existing tech, such as CoAP to expose the local devices as services? Do you intend to just use CoAP or discover the devices via multiple protocols?

Thanks :) Yes, I instead to expose devices as services and I thought that it's also the goal of Akri. I read a nice paper the other day, which explains the same vision: [Combining Cloud and sensors in a smart city environment](Combining Cloud and sensors in a smart city environment). It was written in 2012 when Cloud wsa becoming popular. I have a broader vision which merges cloud, fog and edge computing but the main ideas is the same. If you have some time, give it a read. I think it's worth it, but I'm biased of course 😄

Reply:
I think you are right that these questions are more unique to CoRE. 
> How will developers use CoRE resources in your opinion?

I am certainly not well versed in CoRE. What are some of the main use cases you've seen?

> If you are asking for temperature and brightness, one could not care from which device it comes from. That would allow some virtualization benefits. For instance, we could use the brightness from another device if it fails in the current one. 

You make a good point here about trade offs. 

> If we have an instance/broker per device, then we need to write a new controller which scans the list of device instances to find a result with the temperature sensor as metadata.

This could be an issue even if we have instance per resource. For example, say the Configuration was to find and expose all brightness and temperature resources (maybe a filter of brightness || temperature). Instances would be created for each brightness and each temperature resource, but the name of those instances would be the same: <configuration-name>-digest. So how would Pods request a specific resource? There would need to be a convention of creating separate Configurations for each resource type. 

Maybe we should make the way the Agent names Instances more customizable, such that every instance of a Configuration does not have the same name (besides a digest id). The device plugin name == instance name, so if the instance name is more specific then a resource could be requested with a more specific name. Maybe in the properties map of a DiscoveryResult, there could be a "INSTANCE_NAME" key the Agent could check for. If it exists, the Agent names the instance with the value. Otherwise it gets the default name.

Reply:
> I am certainly not well versed in CoRE. What are some of the main use cases you've seen?

In practise not very much since most works in embedded are close-sourced. But the literature use cases are of any IoT use case since the purpose of CoAP is to be a low-overhead, bandwidth and energy communication protocol yet allowing meaningful RESTful communication. E.g. RFID sensors and healthcare devices are two quite different use-cases I've read about.

One important feature of CoRE is being thought for M2M communication (compared e.g. to udev). That's why it's important for me to support provisioning by resource type, because it requires the least human works. I believe that keeping in mind M2M communication is fundamental when developing for IoT, but it's just my opinion.

In my use case, the cluster must be able to automatically discover resources, assign them to applications or handle availability in case of failure. The applications themselves must be able to work in any cluster with heterogeneous resources so they can only have expectations about the available resource types, nothing about a specific instance. Ideally people develop IoT applications locally and then deploy on a huge variety of edge clusters where Akri instances are running. That will truly bring the cloud to the edge.

If you think about it, in the current industry, a lot of manual repetitive work is done when dealing with IoT. For instance, flashing each embedded device is a nightmare and that's why deployment is always a pain. In the cloud, we can nowadays just push an image and it's done.

I hope this makes sense. I'm open to any criticism since all of you work on Azure and you know better than me.

> Instances would be created for each brightness and each temperature resource, but the name of those instances would be the same: -digest. So how would Pods request a specific resource? There would need to be a convention of creating separate Configurations for each resource type.

Did you mean that the name of those instance would be the same despite being of two different types? And thus how would a Pod request a specific **resource type**. Just clarifying since it's quite easy to get confused with all these similar terms.

> There would need to be a convention of creating separate Configurations for each resource type.

Yup, I thought a bit about it in https://github.com/deislabs/akri/issues/198#issuecomment-758291583. The issue is the potentially huge number of resource types. But maybe in the end it's the best compromise if we suppose that a cluster owner will limit the resource types to a restricted small group.

So far the choices for Configuration are: 1 Configuration x discovery protocol, 1 Configuration x device, 1 Configuration x resource, 1 Configuration per resource type. 

> Maybe we should make the way the Agent names Instances more customizable, such that every instance of a Configuration does not have the same name (besides a digest id). 

I think I understood what you mean, but I'll ask you to provide an example just to be sure 😅

> The device plugin name == instance name, so if the instance name is more specific then a resource could be requested with a more specific name.

That depends on what you mean with "more specific name". Because of the aforementioned M2M goal, a Pod should always just request by resource type like IANA `oic.r.temperature` and never written to need a specific instance like `temperature-digest` IMO.

Reply:
Thanks for sharing the use cases.

> Did you mean that the name of those instance would be the same despite being of two different types? And thus how would a Pod request a specific resource type. 

Yup.

> Maybe we should make the way the Agent names Instances more customizable, such that every instance of a Configuration does not have the same name (besides a digest id).

A DiscoveryHandler (or a call to discover) could return the name that instance should get. So if each resource type is a discovery result, oic.r.temperature could be the returned instance name. That would still have a digest put after it, so ...

>Because of the aforementioned M2M goal, a Pod should always just request by resource type like IANA oic.r.temperature and never written to need a specific instance like temperature-digest IMO.

With Akri, we definitely want to support requesting resources on the Configuration level. so you could request `udev-video` instead of `udev-video-digest`. This brings up an interesting point though: if it isnt 1 Configuration x resource type, that solution would not solve your need to request `oic.r.temperature`

> If you think about it, in the current industry, a lot of manual repetitive work is done when dealing with IoT. For instance, flashing each embedded device is a nightmare and that's why deployment is always a pain. In the cloud, we can nowadays just push an image and it's done.

Definitely agree that there needs to be better device management and hopefully Akri could provide a path to that

Reply:
> This brings up an interesting point though: if it isnt 1 Configuration x resource type, that solution would not solve your need to request oic.r.temperature

Exactly, this is the reason I started to question about the current 1 Configuration x discovery protocol and the consequent additional controller which looks for instances with requested metadata if things are kept as they are.